### ISSUE TITLE
Naive regex matching breaks inlining with comments

### ISSUE THREAD

**Issue by elclanrs** [dherges/ng-packagr#141]
Due to the way the inlining works with regex, when there is a comment that matches the pattern, the build fails. For example, this won't build:

```typescript
/**
@Component({
  selector: 'example',
  templateUrl: './example.component.html'
})
export class Example {
}
*/
@Component({
  selector: 'my-comp',
  template: '<div class="my-comp"></div>'
})
export class MyComponent {
}
```

It will fail with the following error:

> Error: ENOENT: no such file or directory, open '/Users/user/Documents/project/src/app/comp/example.component.html'

Any chance of moving away from gulp-inline-ng2-template and into something like [ts-simple-ast](https://github.com/dsherret/ts-simple-ast) for inlining assets? Working with an AST would avoid this issue altogether.

Perhaps a workaround for now could be to remove comments before inlining.

**Comment by dherges** [comment]
Hi @elclanrs,

oh, yes, using an AST is a good idea!

Maybe [ngast](https://mgechev.github.io/ngast/class/dist-docs/lib/directive-symbol.js~DirectiveSymbol.html#instance-method-getResolvedMetadata) is a good point to read the [CompileTemplateMetadata](https://github.com/angular/angular/blob/57518652288a98eb4f3f81e89c8d82c48b8d2ece/packages/compiler/src/compile_metadata.ts#L243-L250)

Then, we would need to serialize the AST back to TypeScript source. Currently, the inlining happens before the `ngc` compiler creates the `*.metadata.json`, because the AoT metadata references the inlined, transformed CSS styles and HTML templates (after less, scss, ... and so on...)

**Comment by dherges** [comment]
The related issue in gulp-inline-ng2-template: ludohenin/gulp-inline-ng2-template#79 

Challenge is that it's not so easy to make the regexp context-aware.

Problem with the AST might that serialized code differs substantially from the human-written code. already, I don't like copying the `.ts` source files to another location too much ... in build errors, file paths from the internal working directory show up. If we now use the AST, then we wouldn't be able to switch the order of the build steps. Compiling the `.ts` first, then inlining has other issues ... e.g. manipulating the `.metadata.json` and probably sourcemaps could not get lost. The inlining is one of the weak points tbh ...

**Comment by dherges** [comment]
I am thinking loud as in the previous comment ... the other issue in current impl is that the SCSS/HTML pre-processing is hooked into gulp-inline-ng2-template as [style processor](https://github.com/dherges/ng-packagr/blob/master/src/lib/steps/assets.ts#L37)

The more I think of it, the ideal workflow would probably be:

1. In parallel:
   - `ngc` compile `.ts` to `.js` (ES2015) individual files, output in working dir, writes `.d.ts` and `.metadata.json`
   - pre-process `.html` templates, output to working dir
   - pre-process styles `.scss`, `.less`, `.style, ..., output to working dir
2. Inlining:
   1. strip comments in `.js` ES2015 individual files
   2. inline `.html` and `.css` into ES2015 `.js` files
   3. overwrite `.js` files in working dir
3. manipulate `.metadata.json`
4. bundling

**Comment by elclanrs** [comment]
If it helps, this is how we inline assets currently, using the ts-simple-ast library: https://gist.github.com/elclanrs/c1944bf16cb8448b543872995db3ff82

**Comment by dherges** [comment]
Hi @elclanrs,

thank you so much for the code snippet! I only discovered it now!

Are you writing the transformed AST back to a file on disk, then invoke `ngc` from shell/command-line (like a `process.exec("ngc")`)?

**Comment by elclanrs** [comment]
@dherges, yes, exactly. I first git stage all the current changes, then call the script to inline all the assets, then call `ngc`, and finally cleanup the changes.

**Comment by dherges** [comment]
Solved by #279 

---

@elclanrs FYI
